<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Exploratory-Data-Analysis" data-toc-modified-id="Exploratory-Data-Analysis-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Exploratory Data Analysis</a></span><ul class="toc-item"><li><span><a href="#Dataset-shape-and-missing-values" data-toc-modified-id="Dataset-shape-and-missing-values-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>Dataset shape and missing values</a></span><ul class="toc-item"><li><span><a href="#Songs" data-toc-modified-id="Songs-1.1.1"><span class="toc-item-num">1.1.1&nbsp;&nbsp;</span>Songs</a></span></li><li><span><a href="#Users" data-toc-modified-id="Users-1.1.2"><span class="toc-item-num">1.1.2&nbsp;&nbsp;</span>Users</a></span></li></ul></li><li><span><a href="#User's-demopraphics" data-toc-modified-id="User's-demopraphics-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>User's demopraphics</a></span><ul class="toc-item"><li><span><a href="#Gender's-distribution" data-toc-modified-id="Gender's-distribution-1.2.1"><span class="toc-item-num">1.2.1&nbsp;&nbsp;</span>Gender's distribution</a></span></li><li><span><a href="#Age's-distribution" data-toc-modified-id="Age's-distribution-1.2.2"><span class="toc-item-num">1.2.2&nbsp;&nbsp;</span>Age's distribution</a></span></li><li><span><a href="#Country's-distribution" data-toc-modified-id="Country's-distribution-1.2.3"><span class="toc-item-num">1.2.3&nbsp;&nbsp;</span>Country's distribution</a></span></li><li><span><a href="#User's-registration-over-time" data-toc-modified-id="User's-registration-over-time-1.2.4"><span class="toc-item-num">1.2.4&nbsp;&nbsp;</span>User's registration over time</a></span></li><li><span><a href="#User-listening-habits" data-toc-modified-id="User-listening-habits-1.2.5"><span class="toc-item-num">1.2.5&nbsp;&nbsp;</span>User listening habits</a></span><ul class="toc-item"><li><span><a href="#Distribution-of-number-of-songs-per-user" data-toc-modified-id="Distribution-of-number-of-songs-per-user-1.2.5.1"><span class="toc-item-num">1.2.5.1&nbsp;&nbsp;</span>Distribution of number of songs per user</a></span></li><li><span><a href="#Distribution-of-Listening-timestamp" data-toc-modified-id="Distribution-of-Listening-timestamp-1.2.5.2"><span class="toc-item-num">1.2.5.2&nbsp;&nbsp;</span>Distribution of Listening timestamp</a></span></li></ul></li></ul></li><li><span><a href="#Songs-and-Artists-Analysis" data-toc-modified-id="Songs-and-Artists-Analysis-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>Songs and Artists Analysis</a></span><ul class="toc-item"><li><span><a href="#Top-artists-by-user-interaction" data-toc-modified-id="Top-artists-by-user-interaction-1.3.1"><span class="toc-item-num">1.3.1&nbsp;&nbsp;</span>Top artists by user interaction</a></span></li><li><span><a href="#Top-songs-by-user-interactions" data-toc-modified-id="Top-songs-by-user-interactions-1.3.2"><span class="toc-item-num">1.3.2&nbsp;&nbsp;</span>Top songs by user interactions</a></span></li><li><span><a href="#Popularity-of-songs-over-time" data-toc-modified-id="Popularity-of-songs-over-time-1.3.3"><span class="toc-item-num">1.3.3&nbsp;&nbsp;</span>Popularity of songs over time</a></span></li><li><span><a href="#Number-of-songs-per-artists" data-toc-modified-id="Number-of-songs-per-artists-1.3.4"><span class="toc-item-num">1.3.4&nbsp;&nbsp;</span>Number of songs per artists</a></span></li></ul></li></ul></li></ul></div>

In [10]:
# Data manipulation imports
import pandas as pd
import numpy as np

# plotly standard imports
import plotly.graph_objs as go
from plotly.offline import iplot
import matplotlib.pyplot as plt


In [11]:
# Base settings for the plots

# 1. Colors
# Define a color palette
color_palette = {
    'primary': '#C8E4B2',
    'secondary': '#9ED2BE',
    'tertiary': '#7EAA92',
    'quaternary': '#FFD9B7',
    'background': '#fafafa',
    'text': '#333333'
}

# 2. Fonts
font_settings = {
    'family': 'Arial',
    'size': 14,
    'color': color_palette['text']
}

# 3. Layout and Design
layout_settings = {
    'plot_bgcolor': color_palette['background'],
    'paper_bgcolor': color_palette['background'],
    'font': font_settings,
    'margin': go.layout.Margin(l=50, r=50, b=50, t=50)
}

# Sample plot using the defined settings
sample_data = {
    'x': [1, 2, 3, 4],
    'y': [10, 11, 12, 13]
}

In [12]:
# Defines columns name
column_names = ["userid", "timestamp", "artistid", "artist-name", "trackid", "track-name"]

# Load songs dataset
songs = pd.read_csv('/content/drive/MyDrive/Thesis/lastfm-dataset-1K/userid-timestamp-artid-artname-traid-traname.tsv',
                    sep='\t',names=column_names, on_bad_lines ='warn')

# Load users dataset
users = pd.read_csv('/content/drive/MyDrive/Thesis/lastfm-dataset-1K/userid-profile.tsv', sep='\t', on_bad_lines ='warn')

Skipping line 2120260: expected 6 fields, saw 8

Skipping line 2446318: expected 6 fields, saw 8

Skipping line 11141081: expected 6 fields, saw 8

Skipping line 11152099: expected 6 fields, saw 12
Skipping line 11152402: expected 6 fields, saw 8

Skipping line 11882087: expected 6 fields, saw 8

Skipping line 12902539: expected 6 fields, saw 8
Skipping line 12935044: expected 6 fields, saw 8

Skipping line 17589539: expected 6 fields, saw 8



In [13]:
songs.head(10)

userid             timestamp                              artistid  \
0  user_000001  2009-05-04T23:08:57Z  f1b1cf71-bd35-4e99-8624-24a6e15f133a   
1  user_000001  2009-05-04T13:54:10Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
2  user_000001  2009-05-04T13:52:04Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
3  user_000001  2009-05-04T13:42:52Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
4  user_000001  2009-05-04T13:42:11Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
5  user_000001  2009-05-04T13:38:31Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
6  user_000001  2009-05-04T13:33:28Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
7  user_000001  2009-05-04T13:23:45Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
8  user_000001  2009-05-04T13:19:22Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   
9  user_000001  2009-05-04T13:13:38Z  a7f7df4a-77d8-4f12-8acd-5c60c93f4de8   

  artist-name trackid                                  track-name  
0   Deep Dish     NaN  Fuck Me Im Famous (Pacha Ibiza)-09-28-2007  
1        坂本龍一     NaN           Composition 0919 (Live_2009_4_15)  
2        坂本龍一     NaN                        Mc2 (Live_2009_4_15)  
3        坂本龍一     NaN                     Hibari (Live_2009_4_15)  
4        坂本龍一     NaN                        Mc1 (Live_2009_4_15)  
5        坂本龍一     NaN                To Stanford (Live_2009_4_15)  
6        坂本龍一     NaN              Improvisation (Live_2009_4_15)  
7        坂本龍一     NaN                    Glacier (Live_2009_4_15)  
8        坂本龍一     NaN                  Parolibre (Live_2009_4_15)  
9        坂本龍一     NaN             Bibo No Aozora (Live_2009_4_15)

In [14]:
users.head(10)

#id gender   age             country    registered
0  user_000001      m   NaN               Japan  Aug 13, 2006
1  user_000002      f   NaN                Peru  Feb 24, 2006
2  user_000003      m  22.0       United States  Oct 30, 2005
3  user_000004      f   NaN                 NaN  Apr 26, 2006
4  user_000005      m   NaN            Bulgaria  Jun 29, 2006
5  user_000006    NaN  24.0  Russian Federation  May 18, 2006
6  user_000007      f   NaN       United States  Jan 22, 2006
7  user_000008      m  23.0            Slovakia  Sep 28, 2006
8  user_000009      f  19.0       United States  Jan 13, 2007
9  user_000010      m  19.0              Poland   May 4, 2006

# Exploratory Data Analysis

## Dataset shape and missing values

### Songs

In [15]:
# Print number of rows and columns of 'songs' dataset
rows, cols = songs.shape
print(f"The dataset 'songs' has {rows} rows and {cols} columns.")

The dataset 'songs' has 19098853 rows and 6 columns.


In [16]:
# Missing values for each column
missing_values_songs = songs.isnull().sum()
missing_values_songs

userid               0
timestamp            0
artistid        600848
artist-name          0
trackid        2162719
track-name          12
dtype: int64

All the missing values in the 'songs' dataset will be deleted, mainly the id's values for artists and tracks, that will later be used as unique identifiers to the machine learning model.

In [17]:
# Remove missing values on 'songs' dataset
songs.dropna(inplace=True)

In [18]:
# Print number of rows and columns of 'songs' dataset after remove missing values
rows, cols = songs.shape
print(f"The dataset 'songs' has {rows} rows and {cols} columns.")

The dataset 'songs' has 16936134 rows and 6 columns.


In [19]:
# Number of unique values for each column
unique_values_songs = songs.nunique()
unique_values_songs

userid              992
timestamp      15631466
artistid          83905
artist-name       81751
trackid          960402
track-name       693231
dtype: int64

### Users

In [20]:
# Print number of rows and columns of 'users' dataset
rows, cols = users.shape
print(f"The dataset 'users' has {rows} rows and {cols} columns.")

The dataset 'users' has 992 rows and 5 columns.


In [21]:
# Missing values for each column
missing_values_users = users.isnull().sum()
missing_values_users

#id             0
gender        108
age           706
country        85
registered      8
dtype: int64

Different of 'songs' dataset, there is no missing values on id's column. Furthermore, there are some on gender, age, country and registered columns. These missing values will not be deleted, although they will be valuable attributes later on the graph, it is possible to keep some of the nodes on the graph without them.

In [22]:
# Number of unique values for each column
unique_values_users = users.nunique()
unique_values_users

#id           992
gender          2
age            34
country        66
registered    625
dtype: int64

## User's demopraphics

### Gender's distribution

In [23]:
# Distribution of genders
gender_counts = users['gender'].value_counts()
total_users = len(users)

# Calculate percentages
male_percentage = (gender_counts.get('m', 0) / total_users) * 100
female_percentage = (gender_counts.get('f', 0) / total_users) * 100

# Create a Plotly bar chart
bar = go.Bar(x=gender_counts.index, y=gender_counts.values, name='Distribution of Genders',
             marker_color=[color_palette['primary'], color_palette['quaternary']],
             text=gender_counts.values, textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Gender distribution</b>",
    xaxis_title="<b>Gender</b>",
    yaxis_title="<b>Number of Users</b>"
)

# Show the figure
fig.show()

# Print the number and percentage of male and female users
print(f"Number of male users: {gender_counts.get('m', 0)} ({male_percentage:.2f}%)")
print(f"Number of female users: {gender_counts.get('f', 0)} ({female_percentage:.2f}%)")


Number of male users: 502 (50.60%)
Number of female users: 382 (38.51%)


One of the most salient observations that can be gleaned from this dataset pertains to the disparate distribution of genders among the user base. Males hold a marginal majority, accounting for 51.02% of users, while females are not considerably distant, making up 38.82% of the overall user community. Importantly, it should be emphasised that a notable fraction of the user population has opted to leave their gender undeclared. Depending on the extent of its impact, this particular discrepancy could potentially introduce a measure of bias into any subsequent data modelling efforts that rely on gender as a variable. As a result, individuals responsible for data modelling should exercise heightened vigilance to safeguard against any inadvertent biases.

### Age's distribution

In [24]:
# Distribution of age
age_data = users['age'].dropna()

# Calculate histogram bins and frequencies
hist, bin_edges = np.histogram(age_data, bins=50)

# Create x-axis labels (midpoint of each bin)
bin_midpoints = (bin_edges[:-1] + bin_edges[1:]) / 2

# Create a bar chart that mimics a histogram
bar = go.Bar(x=bin_midpoints, y=hist, width=np.diff(bin_edges),
             marker_color=color_palette['primary'],
             text=hist,  # Adding text labels
             textposition='outside')  # Position of text labels

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Age Distribution</b>",
    xaxis_title="<b>Age</b>",
    yaxis_title="<b>Number of Users</b>"
)

# Show the figure
fig.show()



In [25]:
users['age'].describe()

count    286.000000
mean      25.367133
std        8.314952
min        3.000000
25%       21.000000
50%       23.000000
75%       28.000000
max      103.000000
Name: age, dtype: float64

Age, a critical demographic variable, spans a broad spectrum in this dataset, extending from individuals as young as 3 years old to those as senior as 103 years of age. Interestingly, a substantial portion of the dataset, precisely 75%, falls within the age bracket of 21 to 28 years. This noticeable skewness in the data suggests that machine learning models built on this dataset may exhibit a heightened sensitivity to the preferences and behaviours of this specific age demographic. However, this concentration also raises the considerable risk of model overfitting and reduced predictive accuracy for users who fall outside of this age range.

### Country's distribution

In [26]:
# Countries distribution
top_countries = users['country'].value_counts().head(10).sort_values()
bar = go.Bar(
    y=top_countries.index,
    x=top_countries.values,
    orientation='h',
    marker_color=color_palette['primary'],
    text=top_countries.values,
    textposition='outside'
)
fig = go.Figure(data=[bar], layout=layout_settings)
fig.update_layout(
    title="<b>Top 10 Countries by User</b>",
    xaxis_title="<b>Number of Users</b>",
    yaxis_title="<b>Country</b>"
)
fig.show()


In [27]:
# Total number of users
total_users = len(users)

# Number of users from United States and United Kingdom
us_users = len(users[users['country'] == 'United States'])
uk_users = len(users[users['country'] == 'United Kingdom'])

# Calculate percentages
us_percentage = (us_users / total_users) * 100
uk_percentage = (uk_users / total_users) * 100
rest_of_world_percentage = 100 - (us_percentage + uk_percentage)

n_countries = users['country'].nunique()

# Print the results
print(f"Number of countries: {n_countries}")
print(f"Percentage of users from United States: {us_percentage:.2f}%")
print(f"Percentage of users from United Kingdom: {uk_percentage:.2f}%")
print(f"Percentage of users from rest of the world: {rest_of_world_percentage:.2f}%")


Number of countries: 66
Percentage of users from United States: 22.98%
Percentage of users from United Kingdom: 12.70%
Percentage of users from rest of the world: 64.31%


The LastFM dataset draws upon a truly global user base, encompassing individuals from a total of 66 countries. However, it is striking that there is a pronounced emphasis on users from the United States, who constitute 23% of the population, and those from the United Kingdom, who make up an additional 13%. Collectively, these two countries account for a significant 36% of the overall user base. The prevalence of users from these key markets has profound implications for any analytical endeavours based on this dataset.

### User's registration over time

In [28]:
# Convert the 'registered' column to datetime format and coerce any errors
users['registered'] = pd.to_datetime(users['registered'], errors='coerce')

# Remove rows where 'registered' is NaT
users = users.dropna(subset=['registered'])

# Extracting month and year from the 'registered' column
users['month_year'] = users['registered'].dt.to_period('M')

# Grouping by month/year and counting registrations
registration_counts_monthly = users.groupby('month_year').size()

# Create a Plotly line chart
line = go.Scatter(x=registration_counts_monthly.index.astype(str),
                  y=registration_counts_monthly.values,
                  mode='lines+markers',
                  marker_color=color_palette['primary'])

fig = go.Figure(data=[line], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>User Registrations Over Time (Monthly)</b>",
    xaxis_title="<b>Month/Year</b>",
    yaxis_title="<b>Number of Registrations</b>"
)

# Show the figure
fig.show()


<ipython-input-28-b1bcd1657c1c>:8: SettingWithCopyWarning:


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy



In [29]:
# Extracting the year from the 'registered' column
users['year'] = users['registered'].dt.year

# Grouping by year and counting registrations
registration_counts_yearly = users.groupby('year').size()

# Convert the result to a DataFrame for better display
registration_yearly_df = registration_counts_yearly.reset_index()
registration_yearly_df.columns = ['Year', 'Number of Registrations']

# Display the table
display(registration_yearly_df)

Year  Number of Registrations
0  2002                        1
1  2003                       13
2  2004                       77
3  2005                      302
4  2006                      438
5  2007                      147
6  2008                        2
7  2009                        4

The temporal pattern of user registrations in the dataset exhibits a distinctly uneven distribution, providing important clues about periods characterised by heightened activity. Initiated on a modest scale in 2004, the number of user registrations experienced a substantial surge in 2005, culminating in a peak during 2006. This initial burst of momentum, however, started to wane in the year 2007, followed by a marked decrease in registration numbers throughout the subsequent years leading up to 2009 and 2010.

### User listening habits

#### Distribution of number of songs per user

In [30]:
# Distribution of number of songs per user
songs_per_user = songs['userid'].value_counts()

# Calculate histogram bins and frequencies
hist, bin_edges = np.histogram(songs_per_user, bins=30)

# Create x-axis labels (midpoint of each bin)
bin_midpoints = (bin_edges[:-1] + bin_edges[1:]) / 2

# Create a bar chart that mimics a histogram
bar = go.Bar(x=bin_midpoints, y=hist, width=np.diff(bin_edges),
             marker_color=color_palette['primary'],
             text=hist,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Distribution of Number of Songs per User</b>",
    xaxis_title="<b>Number of Songs</b>",
    yaxis_title="<b>Number of Users</b>"
)

# Show the figure
fig.show()

In [31]:
songs['userid'].value_counts().describe()

count       992.000000
mean      17072.715726
std       20800.361269
min           2.000000
25%        3351.500000
50%       10227.000000
75%       22294.000000
max      172042.000000
Name: userid, dtype: float64

The distribution of the number of songs consumed by users exhibits a noticeable positive skew. Most users have interacted with a collection of songs that falls within the range of 3,774 to 25,673. However, there exists a select cohort of users who have demonstrated extraordinarily high levels of engagement, interacting with an astounding 183,103 songs.

#### Distribution of Listening timestamp

In [32]:
# Convert the 'timestamp' column to datetime format
songs['timestamp'] = pd.to_datetime(songs['timestamp'])

# Extract hour from timestamp
songs['hour'] = songs['timestamp'].dt.hour

In [33]:
# Calculate histogram bins and frequencies
hist, bin_edges = np.histogram(songs['hour'], bins=24, range=(0, 24))

# Create x-axis labels (midpoint of each bin)
bin_midpoints = (bin_edges[:-1] + bin_edges[1:]) / 2

# Create a bar chart that mimics a histogram
bar = go.Bar(x=bin_midpoints, y=hist, width=np.diff(bin_edges),
             marker_color=color_palette['primary'],
             text=hist,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Listening hours distribution</b>",
    xaxis_title="<b>Hour of Day</b>",
    yaxis_title="<b>Number of Songs Played</b>"
)

# Show the figure
fig.show()


In [34]:
songs['hour'].describe().round()

count    16936134.0
mean           13.0
std             7.0
min             0.0
25%             8.0
50%            14.0
75%            19.0
max            23.0
Name: hour, dtype: float64

There is a discernible trend concerning the specific periods of the day during which users opt to listen to music. Listening instances begin to escalate noticeably after 8:00 AM, maintaining an upward trajectory until approximately 6:00 PM, when a downward trend ensues. Future investigations might productively explore whether this temporal pattern is consistent across different days of the week or exhibits variations depending on geographical locations.

## Songs and Artists Analysis

### Top artists by user interaction

In [35]:
# Group by 'artistid', count the occurrences, and then sort the results
top_artist_ids = songs.groupby('artistid').size().sort_values(ascending=True).tail(10)

# Replace the 'artistid' index with 'artist-name' for the top artists
top_artist_names = top_artist_ids.index.map(lambda x: songs[songs['artistid'] == x]['artist-name'].iloc[0])
top_artist_ids.index = top_artist_names

# Create a Plotly horizontal bar chart
bar = go.Bar(y=top_artist_ids.index, x=top_artist_ids.values, orientation='h',
             marker_color=color_palette['primary'],
             text=top_artist_ids.values,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Top 10 Artists by User Interactions</b>",
    xaxis_title="<b>Number of Interactions</b>",
    yaxis_title="<b>Artist Name</b>"
)

# Show the figure
fig.show()


In [36]:
# Count the occurrences of each artist in the songs dataset
artist_counts = songs['artist-name'].value_counts()

# Convert the counts to a DataFrame for better display
artist_counts_df = artist_counts.reset_index()
artist_counts_df.columns = ['Artist', 'Number of Times Listened']

# Display the top artists from the table
display(artist_counts_df.head(10))


Artist  Number of Times Listened
0            Radiohead                    111488
1          The Beatles                     98525
2      Nine Inch Nails                     82022
3                 Muse                     60967
4             Coldplay                     60256
5         Depeche Mode                     57802
6  Death Cab For Cutie                     56451
7           Pink Floyd                     55739
8              Placebo                     50632
9        Elliott Smith                     49306

Within the roster of artists featured in the dataset, certain names distinctly rise to prominence. Specifically, Radiohead and The Beatles surface as the most frequently listened-to artists, trailed closely by Nine Inch Nails, Muse, and Coldplay. Intriguingly, although the dataset does not provide explicit genre categorizations, it is noteworthy that these leading artists predominantly align themselves with the rock genre.  

### Top songs by user interactions

In [37]:
# Top songs by user interactions
top_songs = songs['track-name'].value_counts().head(10).sort_values(ascending=True)

# Create a Plotly horizontal bar chart
bar = go.Bar(y=top_songs.index, x=top_songs.values, orientation='h',
             marker_color=color_palette['primary'],
             text=top_songs.values,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Top 10 Songs by User Interactions</b>",
    xaxis_title="<b>Number of Interactions</b>",
    yaxis_title="<b>Song Name</b>"
)

# Show the figure
fig.show()


In [38]:
# Filtering the dataset for the song "Intro"
intro_song = songs[songs['track-name'] == 'Intro']

# 1. Number of unique users who listened to it
unique_users = intro_song['userid'].nunique()

# 2. Number of unique song id
unique_id = intro_song['trackid'].nunique()

# 3. Artists with the song "Intro"
artists_intro = intro_song['artist-name'].nunique()

print(f"Number of unique users who listened to 'Intro': {unique_users}")
print(f"Number of unique song id 'Intro'  : {unique_id}")
print(f"Number of unique artists who performed the song 'Intro': {artists_intro}")

Number of unique users who listened to 'Intro': 751
Number of unique song id 'Intro'  : 1400
Number of unique artists who performed the song 'Intro': 1400


In [39]:
# Filtering the dataset for the song "[Untitled]"
untitled_1_song = songs[songs['track-name'] == '[Untitled]']

# 1. Number of unique users who listened to it
unique_users = untitled_1_song['userid'].nunique()

# 2. Number of unique song id
unique_id = untitled_1_song['trackid'].nunique()

# 3. Artists with the song "[Untitled]"
artists_intro = untitled_1_song['artist-name'].nunique()

print(f"Number of unique users who listened to '[Untitled]': {unique_users}")
print(f"Number of unique song id '[Untitled]'  : {unique_id}")
print(f"Number of unique artists who performed the song '[Untitled]': {artists_intro}")

Number of unique users who listened to '[Untitled]': 451
Number of unique song id '[Untitled]'  : 465
Number of unique artists who performed the song '[Untitled]': 465


In [40]:
# Group by 'trackid', count the occurrences, and then sort the results
top_track_ids = songs.groupby('trackid').size().sort_values(ascending=True).tail(10)

# Replace the 'trackid' index with 'track-name' for the top songs
top_track_names = top_track_ids.index.map(lambda x: songs[songs['trackid'] == x]['track-name'].iloc[0])
top_track_ids.index = top_track_names

# Create a Plotly horizontal bar chart
bar = go.Bar(y=top_track_ids.index, x=top_track_ids.values, orientation='h',
             marker_color=color_palette['primary'],
             text=top_track_ids.values,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Top Songs 10 by User Interactions (filtered)</b>",
    xaxis_title="<b>Number of Interactions</b>",
    yaxis_title="<b>Song Name</b>"
)

# Show the figure
fig.show()

A fascinating discovery within the dataset concerns the conspicuous prominence of generic song titles such as 'Intro' and '[Untitled]'. Upon further scrutiny, 'Intro' is linked to an impressive 1,400 unique artist IDs, while '[Untitled]' is associated with 465. These figures suggest that titles of this nature frequently recur across a variety of artists. Consequently, these two titles were deliberately excluded from the ranking considerations. Reassessing the data utilising unique song IDs leads to a notable alteration in the landscape of top songs. The revised ranking showcases 'Such Great Heights,' 'Love Will Tear Us Apart,' and 'Karma Police' as the most-listened-to songs within the dataset.

### Popularity of songs over time

In [41]:
# Assuming first timestamp as "release date"
song_release_date = songs.groupby('track-name')['timestamp'].min()
songs['song_release_date'] = songs['track-name'].map(song_release_date)
songs['days_since_release'] = (songs['timestamp'] - songs['song_release_date']).dt.days

In [42]:
# Long longevity

# Calculate histogram bins and frequencies
hist, bin_edges = np.histogram(songs['days_since_release'].dropna(), bins=range(0, songs['days_since_release'].max() + 7, 7))

# Create x-axis labels (midpoint of each bin)
bin_midpoints = (bin_edges[:-1] + bin_edges[1:]) / 2

# Create a bar chart that mimics a histogram
bar = go.Bar(x=bin_midpoints, y=hist, width=np.diff(bin_edges),
             marker_color=color_palette['primary'])
fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Longevity of songs popularity</b>",
    xaxis_title="<b>Days Since Release (in bins of 7 days)</b>",
    yaxis_title="<b>Number of Records</b>",
    yaxis_type="log"  # Log scale
)

# Show the figure
fig.show()

In [43]:
# Create bins for days_since_release in intervals of 7 days
bins = pd.cut(songs['days_since_release'], bins=range(0, songs['days_since_release'].max() + 7, 7))

# Count the number of records in each bin
days_since_release_counts = bins.value_counts().sort_index()

# Convert the result to a DataFrame for better display
days_since_release_df = days_since_release_counts.reset_index()
days_since_release_df.columns = ['Days Since Release (Interval)', 'Number of Records']

# Display the table
display(days_since_release_df)


Days Since Release (Interval)  Number of Records
0                          (0, 7]             299476
1                         (7, 14]             175270
2                        (14, 21]             139093
3                        (21, 28]             121020
4                        (28, 35]             112839
..                            ...                ...
391                  (2737, 2744]                  0
392                  (2744, 2751]                  0
393                  (2751, 2758]                  0
394                  (2758, 2765]                  0
395                  (2765, 2772]                  1

[396 rows x 2 columns]

In [44]:
songs['days_since_release'].describe().round()

count    16936134.0
mean          635.0
std           443.0
min             0.0
25%           249.0
50%           603.0
75%           996.0
max          2771.0
Name: days_since_release, dtype: float64

By employing the earliest playback timestamp of each track's unique identifier as a provisional release date, the dataset indicates that songs commonly experience a decrease in popularity following their initial debut week, with a drop of 52% of listening records from the first to the second week. This decline eventually stabilises after roughly 200 days, only to gradually diminish again after approximately 1,500 days, or roughly 4.1 years. It is pertinent to note that the quantity of data registered declines over time, a factor that could potentially affect the trends observed. Nevertheless, the initial popularity trajectory of new songs offers compelling evidence to suggest that users exhibit a marked interest in exploring new musical releases.

### Number of songs per artists

In [45]:
# Artists with most unique songs
artists_song_counts = songs.groupby('artist-name')['trackid'].nunique().sort_values(ascending=True).tail(10)

# Create a Plotly horizontal bar chart
bar = go.Bar(y=artists_song_counts.index, x=artists_song_counts.values, orientation='h',
             marker_color=color_palette['primary'],
             text=artists_song_counts.values,
             textposition='outside')

fig = go.Figure(data=[bar], layout=layout_settings)

# Update specific elements for this plot
fig.update_layout(
    title="<b>Artists with Most Unique Songs</b>",
    xaxis_title="<b>Number of Unique Songs</b>",
    yaxis_title="<b>Artist Name</b>"
)

# Show the figure
fig.show()


Johnny Cash stands out as the most prolific artist in the dataset, contributing an impressive array of nearly 1,000 unique songs. He is closely followed in this distinction by The Beatles, Depeche Mode, Johann Sebastian Bach, and Frédéric Chopin. For those interested in a comprehensive breakdown of artist prolificacy, Figure 9 provides a detailed list.